### ISSUE TITLE
Research needed: OGRE scene managers.

### ISSUE THREAD

**Issue by ohlidalp** [RigsOfRods/rigs-of-rods#1379]
**Warning: Long and rant-ish post ahead! 😆  Try TL;DR below**

Derp! I just discovered that my knowledge on OGRE's scene managers was false, or at least incomplete. I'm not really sure, I have to ask on OGRE forums. 

From my early days with OGRE, I had a faint memory that the most effective general-purpose one is "OctreeSceneManager" and it's also the default, so we're using it in RoR. But I just re-read OGRE's wikipage [SceneManagersFAQ](http://www.ogre3d.org/tikiwiki/tiki-index.php?page=SceneManagersFAQ) and found out we may have not:
* The [intro paragraph](http://www.ogre3d.org/tikiwiki/tiki-index.php?page=SceneManagersFAQ#Selecting_a_Scene_Manager) says the ST_GENERIC scenemanager defaults to OctreeSceneManager plugin if loaded (which RoR does) or dotSceneManager plugin (which doesn't exist anymore so RoR obviously doesn't use it)
* On the same page, the [generic scene manager](http://www.ogre3d.org/tikiwiki/tiki-index.php?page=SceneManagersFAQ#Generic_Scene_Manager_default_) section says that the default manager is built-in and not very effective - programmer should explicitly choose a different one.

Update: Apparently there are 2 methods to create OGRE scenemanager:
* By name specified as string ([link to code](https://bitbucket.org/sinbad/ogre/src/53d901eb7b0c85dacc3f9f8c65bc4f593fdad830/OgreMain/src/OgreSceneManagerEnumerator.cpp?at=v1-9#OgreSceneManagerEnumerator.cpp-145)), such as "OctreeSceneManager". If not found, OGRE throws an exception, which is good.
* Using '[scene type](http://www.ogre3d.org/docs/api/1.9/group___scene.html#ga7383602bd480d43b80c626969b9af914)' mask ([link to code](https://bitbucket.org/sinbad/ogre/src/53d901eb7b0c85dacc3f9f8c65bc4f593fdad830/OgreMain/src/OgreSceneManagerEnumerator.cpp?at=v1-9#OgreSceneManagerEnumerator.cpp-195)). OGRE browses registered plugins and finds one which best matches the entered mask. However, 'ogre.log' (RoR.log in our case) doesn't tell which plugin was selected, which is bad. Further, if no plugin is found, OGRE silently falls back to the default scene manager, which is also bad.

Update2: More complaints towards OGRE:
* You can't tell from docs the actual name of the scenemanager plugin - it's kind of obvious that class OctreeSceneManager will have plugin name "OctreeSceneManager", but IMHO 'obvious' is a no.1 source of misunderstandings - I'd prefer direct docs. [Link to sourcecode](https://bitbucket.org/sinbad/ogre/src/a02bdc5fd2309188dbaa612b638c0881364c4fbd/PlugIns/OctreeSceneManager/src/OgreOctreeSceneManager.cpp?at=default#OgreOctreeSceneManager.cpp-1147)
* You can't tell what 'scene type flags' each scenemanager uses - it's not on the wikipage or in the doxy docs. You need to go to the sourcecode to find that:
     * OctreeSceneManager [ sourcecode](https://bitbucket.org/sinbad/ogre/src/a02bdc5fd2309188dbaa612b638c0881364c4fbd/PlugIns/OctreeSceneManager/src/OgreOctreeSceneManager.cpp?at=default#OgreOctreeSceneManager.cpp-1153)  uses all the flags (FAR, CLOSE, whatever), essentially saying "I'm the best choice for any scene" - so why isn't this the builtin default?
     * PCZSceneManager, described on [wikipage](http://www.ogre3d.org/tikiwiki/tiki-index.php?page=SceneManagersFAQ#Portal_Connected_Zone_Scene_Manager_PCZSM_) as "optimized for indoor scenes" with a note "It's harder to set up because this isn't automatic, you need to place zones and portals" _also uses all flags!_ - see it yourself [in the sources](https://bitbucket.org/sinbad/ogre/src/a02bdc5fd2309188dbaa612b638c0881364c4fbd/PlugIns/OctreeSceneManager/src/OgreOctreeSceneManager.cpp?at=default#OgreOctreeSceneManager.cpp-1153)

## Conclusion (long version)

It turns out I didn't really know how RoR used OGRE all this time. RoR creates it's scene manager as "ST_EXTERIOR_CLOSE" ( [code](https://github.com/RigsOfRods/rigs-of-rods/blob/2523441562c7c352c8fbf81ac85f6cea85756894/source/main/main.cpp#L174) ). In this case, OGRE will list all the available plugins in reverse order of definition in 'Plugins.cfg' and pick the first suitable one. Ever since we're using https://github.com/RigsOfRods/ror-dependencies, we're using 'plugins.cfg' generated by the build, which, by chance, has always "OctreeSceneManager" below "PCZSceneManager". So even though both were enabled in our build, we always ended up with OctreeSceneManager. Whew! 😌 

However, due to a misunderstanding between me and AnotherFoxGuy, the OctreeSceneManager was recently disabled in our dependencies: https://github.com/RigsOfRods/ror-dependencies/commit/5f3589c00012e36c9712fa525a3c68aef7a1459e. Must be fixed ASAP. Origin of the problem: I suggested disabling "OctreeZone" plugin, which is in fact unrelated to "OctreeSceneManager" and actually builds along with "PCZSceneManager" (just to 'avoid confusion' 🙄 ).

## TL; DR

* RoR was using OctreeSceneManager all the time, which is good - but there was some luck involved.
* Recently, building of OctreeSceneManager was accidentally disabled in our dependencies - must be fixed ASAP **Update:** already done in https://github.com/RigsOfRods/ror-dependencies/pull/21
* I've issued a "complaint" to OGRE team: http://www.ogre3d.org/forums/viewtopic.php?f=22&t=93507 😄

**Comment by only-a-ptr** [comment]
I'm removing this from v0.4.8 milestone - the immediate problems are resolved and the rest is a low-priority architectural matter between us and OGRE.

**Comment by only-a-ptr** [comment]
Wow! I've just found a link to my OGRE forum post in the OGRE newsletter: http://www.ogre3d.org/2017/11/07/ogre-ecosystem-roundup. Turns out my suggestion was actually implemented in latest OGRE 1.10.x. Neat!

**Comment by ohlidalp** [comment]
Closing as resolved.
